pkg/ipam/crd.go

```diff
@@ -418,29 +417,31 @@ func (n *nodeStore) updateLocalNodeResource(node *ciliumv2.CiliumNode) {
 		}
 		for _, a := range n.allocators {
 			if a.family == ipFamily {
-				allocator = a
+				markedToReleaseIPMap[ip] = a
 			}
 		}
-		if allocator == nil {
+		if _, ok := markedToReleaseIPMap[ip]; !ok {
 			continue
 		}
+		releaseUpstreamSyncNeeded = true
+	}
+	n.mutex.Unlock()
 
+	for ip, allocator := range markedToReleaseIPMap {
 		// Some functions like crdAllocator.Allocate() acquire lock on allocator first and then on nodeStore.
```

Comment:
This comment should live above with `n.mutex.Unlock()`.

Reply:
thanks, i would adjust the position of the comments